core/src/View/MDSpan/Kokkos_MDSpan_Accessor.hpp

```diff
@@ -212,6 +212,334 @@ struct AtomicAccessorRelaxed {
+  }
+
+  KOKKOS_FUNCTION
+  ReferenceCountedDataHandle(const SharedAllocationTracker& tracker,
+                             pointer data_handle)
+      : m_tracker(tracker), m_handle(data_handle) {}
+
+  // unmanaged ctor
+  template <class OtherElementType,
+            class = std::enable_if_t<std::is_convertible_v<
+                OtherElementType (*)[], value_type (*)[]>>>
+  KOKKOS_FUNCTION ReferenceCountedDataHandle(OtherElementType* ptr)
+      : m_tracker(), m_handle(ptr) {}
+
+  // subview ctor
+  template <class OtherElementType,
+            class = std::enable_if_t<std::is_convertible_v<
+                OtherElementType (*)[], value_type (*)[]>>>
+  KOKKOS_FUNCTION ReferenceCountedDataHandle(
+      const ReferenceCountedDataHandle& other, OtherElementType* ptr)
+      : m_tracker(other.m_tracker), m_handle(ptr) {}
+
+  // converting ctor
+  template <class OtherElementType,
+            class = std::enable_if_t<std::is_convertible_v<
+                OtherElementType (*)[], value_type (*)[]>>>
+  KOKKOS_FUNCTION ReferenceCountedDataHandle(
+      const ReferenceCountedDataHandle<OtherElementType, memory_space>& other)
+      : m_tracker(other.m_tracker), m_handle(other.m_handle) {}
+
+  KOKKOS_DEFAULTED_FUNCTION
+  ReferenceCountedDataHandle(const ReferenceCountedDataHandle&) = default;
+  KOKKOS_DEFAULTED_FUNCTION
+  ReferenceCountedDataHandle(ReferenceCountedDataHandle&&) noexcept = default;
+  KOKKOS_DEFAULTED_FUNCTION
+  ReferenceCountedDataHandle& operator=(const ReferenceCountedDataHandle&) =
+      default;
+  KOKKOS_DEFAULTED_FUNCTION
+  ReferenceCountedDataHandle& operator=(ReferenceCountedDataHandle&&) = default;
+
+  KOKKOS_FUNCTION
+  pointer get() const noexcept { return m_handle; }
+  KOKKOS_FUNCTION
+  explicit operator pointer() const noexcept { return m_handle; }
+
+  bool has_record() const { return m_tracker.has_record(); }
+  auto* get_record() const { return m_tracker.get_record<memory_space>(); }
+  int use_count() const noexcept { return m_tracker.use_count(); }
+
+  std::string get_label() const { return m_tracker.get_label<memory_space>(); }
+  KOKKOS_FUNCTION
+  const SharedAllocationTracker& tracker() const noexcept { return m_tracker; }
+
+  KOKKOS_FUNCTION
+  friend bool operator==(const ReferenceCountedDataHandle& lhs,
+                         const value_type* rhs) {
+    return lhs.m_handle == rhs;
+  }
+
+  KOKKOS_FUNCTION
+  friend bool operator==(const value_type* lhs,
+                         const ReferenceCountedDataHandle& rhs) {
+    return lhs == rhs.m_handle;
+  }
+
+ private:
+  template <class OtherElementType, class OtherSpace>
+  friend class ReferenceCountedDataHandle;
+
+  template <class OtherElementType, class OtherSpace, class NestedAccessor>
+  friend class ReferenceCountedAccessor;
+
+  SharedAllocationTracker m_tracker;
+  pointer m_handle = nullptr;
+};
+
+template <class ElementType>
+class ReferenceCountedDataHandle<ElementType, AnonymousSpace> {
```

Comment:
It's not clear what gets specialized. I'd recommend to either add some comment on the specialized things, or if only a very tiny subset of things are specialized, probably you could do some SFINAE stuff in the primary class declaration ?

Reply:
I tried that but it was somewhat messy. The specialization piece has to do with convertibility i.e. it's all in the constructors. I personally felt that the specialization tack was cleaner.

Reply:
I will add some comments.

Reply:
Speaking of code duplication...

Reply:
I still think that it's bad to specialize the whole class if you have a single difference (the conversion constructor).

Reply:
If I remember correct there was some sort of issue with a combination of ambiguity and round-trip via conversion to and construction from `default_accessor` where we couldn't figure out how to prevent conversions we did NOT want to allow doing it in a single class. I can try again and see if I can figure it out.

Reply:
I will try again and if I reencounter the real issue report it in an understandable way.

Reply:
I don't think making the SFINAE on the converting constructor significantly more complex is easier to read than two specializations. If y'all want two specializations we should inherit from a common type that provides most of the functionality. But given how trivial the repeated code is I don't know if it's worth it.

There's also the issue that @crtrott mentioned with round trip conversions but we wrote this code a month or so ago so I don't quite remember what it was. I can look into it.